Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -172,4 +210,129 @@ end CommutativeLemmas
+
+variable [AddMonoid A] [CovariantClass A A (· + ·) (· ≤ ·)]
+  [CovariantClass A A (Function.swap (· + ·)) (· ≤ ·)]
+
+variable (f g : AddMonoidAlgebra R A)
+
+theorem supDegree_mul_le : (f * g).supDegree ≤ f.supDegree + g.supDegree :=
+  sup_support_mul_le (fun {_ _} => (WithBot.coe_add _ _).le) f g
+#align add_monoid_algebra.sup_degree_mul_le AddMonoidAlgebra.supDegree_mul_le
+
+end SupDegree
+
+section InfDegree
+
+variable [SemilatticeInf A]
+
+/-- Let `R` be a semiring, let `A` be a `SemilatticeInf`, and let `f : R[A]` be an
+an element of `AddMonoidAlgebra R A`.  The inf-degree of `f` takes values in `WithTop A`
+and it is the infimum of the support of `f` or `⊤`, depending on whether `f` is non-zero or not.
+If `A` has a linear order, then this notion coincides with the usual one, using the minimum of
+the exponents. -/
+@[reducible]
+def infDegree (f : AddMonoidAlgebra R A) : WithTop A :=
+  f.support.inf (↑)
+#align add_monoid_algebra.inf_degree AddMonoidAlgebra.infDegree
+
+theorem le_infDegree_add (f g : AddMonoidAlgebra R A) :
+    f.infDegree ⊓ g.infDegree ≤ (f + g).infDegree :=
+  --  Porting note: the coercion now needs the explicit type ascription
+  sup_support_add_le ((↑) : Aᵒᵈ → WithBot Aᵒᵈ) f g
+#align add_monoid_algebra.le_inf_degree_add AddMonoidAlgebra.le_infDegree_add
+
+variable [AddMonoid A] [CovariantClass A A (· + ·) (· ≤ ·)]
+  [CovariantClass A A (Function.swap (· + ·)) (· ≤ ·)] (f g : AddMonoidAlgebra R A)
+
+theorem le_infDegree_mul : f.infDegree + g.infDegree ≤ (f * g).infDegree :=
+  sup_support_mul_le (fun {_ _ : Aᵒᵈ} => (WithBot.coe_add _ _).le) _ _
+#align add_monoid_algebra.le_inf_degree_mul AddMonoidAlgebra.le_infDegree_mul
+
+end InfDegree
+
+section MaxDegree
+
+variable [AddZeroClass A] [LinearOrderedAddCommMonoid B] [OrderBot B] (D : A →+ B)
+
+/-- Let `R` be a semiring, let `A` be a `SemilatticeSup`, and let `f : R[A]` be an
+an element of `AddMonoidAlgebra R A`.  The max-degree of `f` takes values in `WithBot A`
+and it is the supremum of the support of `f` or `⊥`, depending on whether `f` is non-zero or not.
+If `A` has a linear order, then this notion coincides with the usual one, using the maximum of
+the exponents. -/
+@[reducible]
+def maxDegree (f : AddMonoidAlgebra R A) : B :=
+  f.support.sup D
+#align add_monoid_algebra.max_degree AddMonoidAlgebra.maxDegree
+
+theorem maxDegree_add_le (f g : AddMonoidAlgebra R A) :
+    (f + g).maxDegree D ≤ max (f.maxDegree D) (g.maxDegree D) :=
+  sup_support_add_le D f g
+#align add_monoid_algebra.max_degree_add_le AddMonoidAlgebra.maxDegree_add_le
+
+theorem maxDegree_mul_le (f g : AddMonoidAlgebra R A) :
+    (f * g).maxDegree D ≤ f.maxDegree D + g.maxDegree D :=
+  sup_support_mul_le (fun {_ _} => (AddMonoidHom.map_add D _ _).le) f g
+#align add_monoid_algebra.max_degree_mul_le AddMonoidAlgebra.maxDegree_mul_le
+
+end MaxDegree
+
+section MinDegree
+
+variable [AddZeroClass A] [LinearOrderedAddCommMonoid T] [OrderTop T] (D : A →+ T)
```

Comment:
Is there a reason not to generalize this to
```suggestion
variable [AddZeroClass A] [LinearOrderedAddCommMonoid T] [OrderTop T] (D : A → T)
```

Reply:
Ah, I found https://github.com/leanprover-community/mathlib/pull/15296#issuecomment-1185746141:

> I'm not against generality; in fact, I just found one use case in mathlib where `B` is not a linear_order: [mv_polynomial.degrees](https://leanprover-community.github.io/mathlib_docs/data/mv_polynomial/variables.html#mv_polynomial.degrees), so I'm in favor of keeping the smilattice_inf versions (`D` is still a homomorphism though.)
> 
> I do think the current lemmas would be more convenient to use if specialized to the linear_order case. If they're gonna be used in 3 situations that all have `B` being linear_order then we should probably do the specialization here instead at each of the 3 situations. Maybe we should reserve `max_degree` and `min_degree` for the linear_order case, and rename them to `sup_degree` and `inf_degree` in the semilattice_sup/inf case. The linear_order version would be stated using `[linear_ordered_add_comm_monoid B]` and `D : A →+ B`, which would be more convenient to use, as the D0, Dm hypotheses cannot be supplied by typeclass inference.
> 
> There are also many lemmas that hold for linear_order but not general semilattice_inf; [polynomial.degree_add_eq_left_of_degree_lt](https://leanprover-community.github.io/mathlib_docs/data/polynomial/degree/definitions.html#polynomial.degree_add_eq_left_of_degree_lt) (consider p = xy^2-yx^2 and q = yx^2), and ~[polynomial.degree_mul'](https://leanprover-community.github.io/mathlib_docs/data/polynomial/degree/definitions.html#polynomial.degree_mul') (leading_coeff only really makes sense in linear_order case)~ (sorry, can't really state it this way unless D is injective; would need to use [graded_ring.proj](https://leanprover-community.github.io/mathlib_docs/ring_theory/graded_algebra/basic.html#graded_ring.proj)/decompose).

Reply:
I am in favour of switching back to `SemilatticeInf`, but am not opposed to leaving it as is, given the controversial nature of the assumption!

In other situations, I remember regretting not having made a possible generalisation.  I think that I never regretted generalising too much!

Reply:
Ok, let's switch back to the general version.